pkg/api/helpers/download.go

```diff
@@ -0,0 +1,208 @@
+package helpers
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/go-openapi/swag"
+	"github.com/treeverse/lakefs/pkg/api/apigen"
+	"github.com/treeverse/lakefs/pkg/uri"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	MinDownloadPartSize        int64 = 1024 * 64       // 64KB
+	DefaultDownloadPartSize    int64 = 1024 * 1024 * 8 // 8MB
+	DefaultDownloadConcurrency       = 10
+)
+
+type Downloader struct {
+	Client     *apigen.ClientWithResponses
+	PreSign    bool
+	HTTPClient *http.Client
+	PartSize   int64
+}
+
+type downloadPart struct {
+	Number     int
+	RangeStart int64
+	PartSize   int64
+}
+
+func NewDownloader(client *apigen.ClientWithResponses, preSign bool) *Downloader {
+	// setup http client
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.MaxIdleConnsPerHost = 10
+	httpClient := &http.Client{
+		Transport: transport,
+	}
+
+	return &Downloader{
+		Client:     client,
+		PreSign:    preSign,
+		HTTPClient: httpClient,
+		PartSize:   DefaultDownloadPartSize,
+	}
+}
+
+// Download downloads an object from lakeFS to a local file, create the destination directory if needed.
+func (d *Downloader) Download(ctx context.Context, src uri.URI, dst string) error {
+	// create destination dir if needed
+	dir := filepath.Dir(dst)
+	_ = os.MkdirAll(dir, os.ModePerm)
+
+	// download object
+	var err error
+	if d.PreSign {
+		// download using presigned multipart download, it will fall back to presign single object download if needed
+		err = d.downloadPresignMultipart(ctx, src, dst)
+	} else {
+		err = d.downloadObject(ctx, src, dst)
+	}
+	if err != nil {
+		return fmt.Errorf("download failed: %w", err)
+	}
+	return nil
+}
+
+func (d *Downloader) downloadPresignMultipart(ctx context.Context, src uri.URI, dst string) (err error) {
+	// get object metadata for size and physical address (presigned)
+	statResp, err := d.Client.StatObjectWithResponse(ctx, src.Repository, src.Ref, &apigen.StatObjectParams{
+		Path:    *src.Path,
+		Presign: swag.Bool(true),
+	})
+	if err != nil {
+		return err
+	}
+
+	// fallback to download if missing size
+	if statResp.JSON200 == nil || statResp.JSON200.SizeBytes == nil {
+		return d.downloadObject(ctx, src, dst)
+	}
+
+	// check if the object is small enough to download in one request
+	sizeBytes := *statResp.JSON200.SizeBytes
+	if sizeBytes < d.PartSize {
+		return d.downloadObject(ctx, src, dst)
+	}
```

Comment:
```suggestion
	// fallback to download if missing size
	if statResp.JSON200 == nil || statResp.JSON200.SizeBytes == nil || *statResp.JSON200.SizeBytes < d.PartSize {
		return d.downloadObject(ctx, src, dst)
	}
```
And maybe add a log to indicate why we fallback to download